### PR TITLE
[pruning][core][feature] Implement prune for structured pruning

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89777
* #90315

Summary:

This PR implements `prune` in BaseStructuredSparsifier:

`prune` is a function that takes in a model with structured sparsity parametritizations (the result of `prepare`) and will return a resized model with the masked out weights removed.

`prune` is defined by a mapping from **patterns** to different **pruning functions**.
	- **patterns** are just sequences of operations, for example `(nn.Linear, activation, nn.Linear)`
	- **pruning functions** are functions that take in an matched pattern as args and will resize the appropriate layer sizes and weights.
	  ```
	  def prune_linear_activation_linear(linear1, activation, linear2):
		pass
	  ```
	- This is one line in the pattern config `(nn.Linear, activation, nn.Linear): prune_linear_activation_linear`

At a high level `prune` works by finding instances of the graph that match different patterns and then calling the mapped pruning functions on those matched patterns.
This is unlike the previous code which attempted to do both at the same time.

There may be some gaps in the patterns compared to the previous implementation, but the conversion functionality support should be the same.

Currently we have pruning functions for the following patterns:
    - linear -> linear
    - linear -> activation -> linear
    - conv2d -> conv2d
    - conv2d -> activation -> conv2d
    - conv2d -> activation -> pool -> conv2d
    - conv2d -> pool -> activation -> conv2d
    - conv2d -> adaptive pool -> flatten -> linear

Added in MyPy type hints as well for the prune_functions.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/pruning/_experimental/pruner/prune_functions.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+    "prune_conv2d",
+    "prune_conv2d_conv2d",
+    "prune_conv2d_activation_conv2d",
+    "prune_conv2d_activation_pool_conv2d",
+    "prune_conv2d_pool_activation_conv2d",
+    "prune_conv2d_pool_flatten_linear",
+]
+
+# BIAS PROPOGATION
+def _get_adjusted_next_layer_bias(next_layer, pruned_biases, mask):
```
Comment (vkuzo) [inline torch/ao/pruning/_experimental/pruner/prune_functions.py:22]: can we add mypy types for all the functions in this file?

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 2]
```diff
@@ -56,18 +165,60 @@ def _prepare(self, *args, **kwargs) -> None:
...
+
+        self.traced = symbolic_trace(self.model)
+        modules = dict(self.traced.named_modules())
+
+        # Right now we check for matches simply by iterating across all the patterns
+        # if this is slow we can store patterns in a trie-structure and modify this code for faster lookup
+
+        for node in self.traced.graph.nodes:
+            for pattern, convert_fn in self.patterns.items():
+                matched = apply_match(modules, pattern, node, [])
```
Comment (vkuzo) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:199]: is this function from the previous PR?  Would it make more sense to put it in this PR?

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 3]
```diff
@@ -56,18 +165,60 @@ def _prepare(self, *args, **kwargs) -> None:
...
+                ):
+                    convert_block = []
+                    for node in matched:
+                        if node.op == "call_module":
+                            convert_block.append(modules.get(node.target))
+                        elif node.op == "call_function":
+                            convert_block.append(node.target)
+                    convert_fn(*convert_block)
+
+        # remove bias hooks, since biases are propogated during module conversion.
```
Comment (vkuzo) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:218]: can this be done during module conversion instead?  Either in `convert_fn` or directly after it.

**Review by vkuzo — COMMENTED** [review]
Overall looks good, is there any testing that needs to happen?

**Comment by vkuzo** [comment]
> is there any testing that needs to happen?

ignore me, I see the test plan now

**Review by vkuzo — APPROVED** [review]

File: torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py [hunk 4]
```diff
@@ -27,16 +131,21 @@ class BaseStructuredSparsifier(BaseSparsifier):
...
         super().__init__(defaults)
+        if patterns is None:
+            patterns = _get_default_structured_pruning_patterns()
+        self.patterns = patterns
 
     def make_config_from_model(
         self,
         model: nn.Module,
-        SUPPORTED_MODULES: Set[Type] = SUPPORTED_STRUCTURED_PRUNING_MODULES,
+        SUPPORTED_MODULES: Set[Type] = None,
```
Comment (z-a-f) [inline torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:143]: qq: If you set it to `None`, should it be `Optional[Set]`, or is the linter ok with this?

Comment (jcaip) [reply torch/ao/pruning/_experimental/pruner/base_structured_sparsifier.py:143]: The linter was fine with this, but yeah it should be `Optional`

**Review by vkuzo — APPROVED** [review]
